File: src/main/scala/com/digitalasset/testing/ledger/clock/SystemTimeProvider.scala
/*
 * Copyright 2019 Digital Asset (Switzerland) GmbH and/or its affiliates
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package com.digitalasset.testing.ledger.clock
import java.time.{Clock, Instant}

import io.grpc.ManagedChannel

object SystemTimeProviderFactory extends TimeProviderFactory {
  override def getTimeProvider(channel: ManagedChannel,
                               ledgerId: String): TimeProvider =
    new SystemTimeProvider()
}

class SystemTimeProvider extends TimeProvider {
  private val systemUtcClock = Clock.systemUTC()
  override def getCurrentTime: Instant = systemUtcClock.instant()

  def setCurrentTime(time: Instant): Unit =
    throw new UnsupportedOperationException(
      "System Time Provider: System time cannot be changed.")
}

File: src/main/java/com/digitalasset/testing/ledger/clock/SystemTimeProvider.java
/*
 * Copyright 2019 Digital Asset (Switzerland) GmbH and/or its affiliates
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package com.digitalasset.testing.ledger.clock;

import java.time.Clock;
import java.time.Instant;
import java.util.function.Supplier;

public class SystemTimeProvider implements TimeProvider {
  public static Supplier<TimeProvider> factory() {
    return SystemTimeProvider::new;
  }

  private static final Clock systemUtcClock = Clock.systemUTC();

  @Override
  public Instant getCurrentTime() {
    return systemUtcClock.instant();
  }

  @Override
  public void setCurrentTime(Instant time) {
    throw new UnsupportedOperationException("System Time Provider: System time cannot be changed.");
  }
}
